crates/iceberg-catalog/src/api/management/v1/warehouse.rs

```diff
@@ -75,6 +151,305 @@ pub trait WarehouseService<C: Catalog, A: AuthZHandler, S: SecretStore> {
             warehouse_id: warehouse_id.into_uuid(),
         })
     }
+
+    async fn list_projects(
+        context: ApiContext<State<A, C, S>>,
+        request_metadata: RequestMetadata,
+    ) -> Result<ListProjectsResponse> {
+        // ------------------- AuthZ -------------------
+        let projects = A::check_list_projects(&request_metadata, context.v1_state.auth).await?;
+
+        // ------------------- Business Logic -------------------
+        if let Some(projects) = projects {
+            return Ok(ListProjectsResponse {
+                project_ids: projects.into_iter().map(|id| id.into_uuid()).collect(),
+            });
+        }
+
+        let projects = C::list_projects(context.v1_state.catalog).await?;
+        Ok(ListProjectsResponse {
+            project_ids: projects.into_iter().map(|id| id.into_uuid()).collect(),
+        })
+    }
+
+    async fn list_warehouses(
+        request: ListWarehouseRequest,
+        context: ApiContext<State<A, C, S>>,
+        request_metadata: RequestMetadata,
+    ) -> Result<ListWarehouseResponse> {
+        // ------------------- AuthZ -------------------
+        let project_id = ProjectIdent::from(
+            request.project_id.ok_or(
```

Comment:
Are we using `ListWarehouseRequest` elsewhere? We could make the field required on the object already

Reply:
I did it this way only to improve the error a bit

Reply:
I was thinking about the openapi rep, is it shown as optional there too?